### PR TITLE
add resize for hgraph

### PR DESCRIPTION
**Pull request by LHT129**
- visitlist for hgraph is fixed, now make it flexible
- fix typo error on graph interface

### CODE REVIEW

**Review by inabao — APPROVED** [review]
LGTM

File: src/index/hgraph_index.cpp [hunk 1]
```diff
@@ -671,5 +673,14 @@ HGraphIndex::calc_distance_by_id(const float* vector, int64_t id) const {
         }
     }
 }
+void
+HGraphIndex::resize(uint64_t new_size) {
+    auto cur_size = this->bottom_graph_->MaxCapacity();
+    if (new_size > cur_size) {
+        vsag::Vector<std::shared_mutex>(new_size, allocator_).swap(this->neighbors_mutex_);
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp:700]: Is this thread-safe?

File: src/index/hgraph_index.cpp [hunk 2]
```diff
@@ -671,5 +673,14 @@ HGraphIndex::calc_distance_by_id(const float* vector, int64_t id) const {
...
+        pool_ = std::make_shared<hnswlib::VisitedListPool>(1, new_size, allocator_);
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp]: Need copy original data to new `pool_`

File: src/index/hgraph_index.cpp (hunk 1 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp:700]: no concurrent in resize

File: src/index/hgraph_index.cpp (hunk 2 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp]: no need to copy, the original visitlist is discarded

File: src/index/hgraph_index.cpp [hunk 3]
```diff
@@ -209,7 +211,7 @@ HGraphIndex::hnsw_add(const DatasetPtr& data) {
     auto* ids = data->GetIds();
     auto* datas = data->GetFloat32Vectors();
     auto cur_count = this->bottom_graph_->TotalCount();
-    vsag::Vector<std::shared_mutex>(total + cur_count, allocator_).swap(this->neighbors_mutex_);
+    this->resize(total + cur_count);
```
Comment (wxyucs) [inline src/index/hgraph_index.cpp:239]: is it ok if we call hnsw_add frequently

Comment (LHT129) [reply src/index/hgraph_index.cpp:239]: we need a strategy for resize, less than 1000 increase 100, more than 1000 increase 1000? like this

File: src/index/hgraph_index.cpp (hunk 1 above)
Comment (jiaweizone) [reply src/index/hgraph_index.cpp:700]: maybe other thread is  lock `neighbors_mutex_` ?

Comment (LHT129) [reply src/index/hgraph_index.cpp:700]: yeah, lock neighbors

Comment (jiaweizone) [reply src/index/hgraph_index.cpp:700]: can simply write to `this->neighbors_mutex_.resize(new_size)` ?

**Review by jiaweizone — APPROVED** [review]
LGTM

**Pull request merged**